java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java

```diff
@@ -75,11 +76,33 @@ protected void onMethodInvocationFound(MethodInvocationTree tree) {
     // can be null when using raw types
     Type collectionParameterType = getTypeParameter(collectionType);
 
-    if (collectionParameterType != null && !collectionParameterType.isUnknown() && !isArgumentCompatible(argumentType, collectionParameterType)) {
+    if (collectionParameterType != null && !collectionParameterType.isUnknown() && !isArgumentCompatible(argumentType, collectionParameterType)
+      && !(useTypeParameters(argumentType) || useTypeParameters(collectionType))) {
       reportIssue(MethodsHelper.methodName(tree), MessageFormat.format("A \"{0}<{1}>\" cannot contain a \"{2}\"", collectionType, collectionParameterType, argumentType));
     }
   }
 
+  // FIXME SONARJAVA-1298 - method to remove
+  // should correctly resolve type parameters, and consequently correctly handle argument compatibility
+  private static boolean useTypeParameters(Type type) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "useTypeParameters" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)